canals/pipeline/save_load.py

```diff
@@ -67,97 +69,163 @@ def marshal_pipelines(pipelines: Dict[str, Pipeline]) -> Dict[str, Any]:
+                if receiver[0] == old_name:
+                    receiver[0] = receiver[0].replace(old_name, new_name)
+                connection["sender"] = ".".join(sender)
+                connection["receiver"] = ".".join(receiver)
+
+
+def _remove_duplicate_instances(components: Dict[str, Any]):
+    """
+    Remove duplicate declaration of the same component instance.
+    If two or more components have the same hash remove all duplicates
+    and point them to a single component.
+
+    A structure like this:
+    ```
+    {
+        "comp1": {"hash": 123, ...},
+        "comp2": {"hash": 123, ...},
+        "comp3": {"hash": 456, ...}
+    }
+    ```
+
+    will become:
+    ```
+    {
+        "comp1": {"hash": 123, ...},
+        "comp2": "comp1",
+        "comp3": {"hash": 456, ...}
+    }
+    ```
     """
-    Loads the content of a schema generated by `marshal_pipelines()` into
+    for comp_name, comp in components.items():
+        if isinstance(comp, str):
+            # This component is just a pointer to another one
+            continue
+        for other_comp_name, other_comp in components.items():
+            if other_comp_name == comp_name:
+                # It's the same component, skip it
+                continue
+            if isinstance(other_comp, str):
+                # This component is just a pointer to another one
+                continue
+            if other_comp["hash"] == comp["hash"]:
+                components[other_comp_name] = comp_name
+
+
+def _cleanup_marshalled_data(data: Dict[str, Any]):
+    # Delete components declared in each pipeline, connections are enough
+    # since we're declaring components globally
+    for pipe in data["pipelines"].values():
+        del pipe["components"]
+
+    # Delete components hashes, we don't need them anymore
+    for comp in data["components"].values():
+        if isinstance(comp, dict):
+            del comp["hash"]
+
+
+def unmarshal_pipelines(data: Dict[str, Any]) -> Dict[str, Pipeline]:
+    """
+    Loads the content of a dictionary generated by `marshal_pipelines()` into
     a dictionary of named Pipelines.
 
     Args:
-        schema: the schema of the pipelines, as generated by `marshal_pipelines()`.
+        data: pipelines' data, as generated by `marshal_pipelines()`.
 
     Returns:
         The pipelines as a dictionary of `{"pipeline-name": <pipeline object>}`.
+    """
 
-    Raises PipelineUnmarshalError: if any Component class has not been imported before loading.
+    components: Dict[str, Component] = {}
+    # Create a mapping of component names and component instances
+    for name, comp_data in data["components"].items():
+        if isinstance(comp_data, str):
+            # This component refers to another one
+            components[name] = components[comp_data]
+            continue
+        init_params = comp_data["init_parameters"]
+        components[name] = component.registry[comp_data["type"]](**init_params)
```

Comment:
Shouldn't use the component's `from_dict()` method here?

```suggestion
    for name, comp_data in data["components"].items():
        if isinstance(comp_data, str):
            # This component refers to another one
            components[name] = components[comp_data]
        else:
            components[name]=component.registry[comp_data["type"]].from_dict(comp_data)
```